crypto/x509/v3_purp.c

```diff
@@ -471,9 +472,17 @@ static void x509v3_cache_extensions(X509 *x)
     if (!X509_NAME_cmp(X509_get_subject_name(x), X509_get_issuer_name(x))) {
         x->ex_flags |= EXFLAG_SI;
         /* If SKID matches AKID also indicate self signed */
-        if (X509_check_akid(x, x->akid) == X509_V_OK &&
-            !ku_reject(x, KU_KEY_CERT_SIGN))
-            x->ex_flags |= EXFLAG_SS;
+        if (X509_check_akid(x, x->akid) == X509_V_OK) {
+            /* check if the signature alg matches the PUBKEY alg */
+            EVP_PKEY *pkey = X509_get0_pubkey(x);
+            X509_ALGOR *s_algor = &x->cert_info.signature;
+            int s_pknid = NID_undef, s_mdnid = NID_undef;
+            if (pkey != NULL
+                    && OBJ_find_sigid_algs(OBJ_obj2nid(s_algor->algorithm),
+                                           &s_mdnid, &s_pknid)
```

Comment:
I think you don't need s_mdnid here, just passing NULL for that parameter should be enough

Reply:
Thanks for the hint - I've removed it.